### PR TITLE
add system variable tidb_opt_prefix_index_single_scan

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What is changed, added or deleted? (Required)

Add system variable `tidb_opt_prefix_index_single_scan`.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.4 (TiDB 6.4 versions)
- [ ] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by songrijie** [comment]
adjust the variable name `tidb_opt_prefer_prefix_index_single_scan` to match prefix `tidb_opt_`. 

rest LGTM

**Comment by xuyifangreeneyes** [comment]
> adjust the variable name `tidb_opt_prefer_prefix_index_single_scan` to match prefix `tidb_opt_`.
> 
> rest LGTM

Rename to `tidb_opt_prefix_index_single_scan` for brevity.

File: system-variables.md [hunk 1]
```diff
@@ -2364,6 +2364,56 @@ explain select * from t where age=5;
 - 范围：`[0, 1]`
 - 这个变量用来控制是否在 profile 输出中标记出对应的 SQL 语句，用于定位和排查性能问题。
 
+### `tidb_opt_prefix_index_single_scan` <span class="version-mark">从 v6.4 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`ON`
+- 这个变量用于控制是否应用一些优化使得使用前缀索引时避免不必要的回表。
```
Comment (TomShawn) [inline system-variables.md:2372]:
```suggestion
- 这个变量用于控制 TiDB 优化器是否通过应用优化，在查询中使用前缀索引时避免不必要的回表，从而提高查询性能。
```

File: system-variables.md [hunk 2]
```diff
@@ -2364,6 +2364,56 @@ explain select * from t where age=5;
...
+- 将该变量设置为 `ON` 时，假设 `col` 列是索引前缀列，`col is null` 或者 `col is not null` 会被归为索引上的过滤条件而不是回表时的过滤条件，从而避免不必要的回表。
```
Comment (TomShawn) [inline system-variables.md:2373]:
```suggestion
- 该变量值为 `ON` 时，开启优化。此时，假设一张表中 `col` 列是索引前缀列，查询语句中的 `col is null` 或者 `col is not null` 条件会被归为索引上的过滤条件，而不是回表时的过滤条件，从而避免不必要的回表。
```

File: system-variables.md [hunk 3]
```diff
@@ -2364,6 +2364,56 @@ explain select * from t where age=5;
...
+
+创建一张带前缀索引的表。
+```
+mysql> CREATE TABLE t (a INT, b VARCHAR(10), c INT, INDEX idx_a_b(a, b(5)));
+Query OK, 0 rows affected (0.02 sec)
+```
```
Comment (TomShawn) [inline system-variables.md:2379]:
````suggestion
该变量的使用示例如下。

创建一张带前缀索引的表：

```sql
CREATE TABLE t (a INT, b VARCHAR(10), c INT, INDEX idx_a_b(a, b(5)));
```
````

File: system-variables.md [hunk 4]
```diff
@@ -2364,6 +2364,56 @@ explain select * from t where age=5;
...
++-------------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
+| HashAgg                       | 1.00    | root      |                              | funcs:count(Column#8)->Column#5                       |
+| └─IndexLookUp                 | 1.00    | root      |                              |                                                       |
+|   ├─IndexRangeScan(Build)     | 99.90   | cop[tikv] | table:t, index:idx_a_b(a, b) | range:[1 -inf,1 +inf], keep order:false, stats:pseudo |
+|   └─HashAgg(Probe)            | 1.00    | cop[tikv] |                              | funcs:count(1)->Column#8                              |
+|     └─Selection               | 99.90   | cop[tikv] |                              | not(isnull(test.t.b))                                 |
+|       └─TableRowIDScan        | 99.90   | cop[tikv] | table:t                      | keep order:false, stats:pseudo                        |
++-------------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
```
Comment (TomShawn) [inline system-variables.md:2398]:
````suggestion
此时关闭 `tidb_opt_prefix_index_single_scan`：

```sql
SET tidb_opt_prefix_index_single_scan = 'OFF';
```

对于以下使用前缀索引的查询，从执行计划结果中可见，执行使用了索引 `idx_a_b` 但需要回表（出现了 `TableRowIDScan` 算子）。

```sql
EXPLAIN FORMAT='brief' SELECT COUNT(1) FROM t WHERE a = 1 AND b IS NOT NULL;

+-------------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
| id                            | estRows | task      | access object                | operator info                                         |
+-------------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
| HashAgg                       | 1.00    | root      |                              | funcs:count(Column#8)->Column#5                       |
| └─IndexLookUp                 | 1.00    | root      |                              |                                                       |
|   ├─IndexRangeScan(Build)     | 99.90   | cop[tikv] | table:t, index:idx_a_b(a, b) | range:[1 -inf,1 +inf], keep order:false, stats:pseudo |
|   └─HashAgg(Probe)            | 1.00    | cop[tikv] |                              | funcs:count(1)->Column#8                              |
|     └─Selection               | 99.90   | cop[tikv] |                              | not(isnull(test.t.b))                                 |
|       └─TableRowIDScan        | 99.90   | cop[tikv] | table:t                      | keep order:false, stats:pseudo                        |
+-------------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
6 rows in set (0.00 sec)
```
````

File: system-variables.md [hunk 5]
```diff
@@ -2364,6 +2364,56 @@ explain select * from t where age=5;
...
++--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
+| id                       | estRows | task      | access object                | operator info                                         |
++--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
+| StreamAgg                | 1.00    | root      |                              | funcs:count(Column#7)->Column#5                       |
+| └─IndexReader            | 1.00    | root      |                              | index:StreamAgg                                       |
+|   └─StreamAgg            | 1.00    | cop[tikv] |                              | funcs:count(1)->Column#7                              |
+|     └─IndexRangeScan     | 99.90   | cop[tikv] | table:t, index:idx_a_b(a, b) | range:[1 -inf,1 +inf], keep order:false, stats:pseudo |
++--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
+4 rows in set (0.00 sec)
+```
```
Comment (TomShawn) [inline system-variables.md:2415]:
````suggestion
此时打开 `tidb_opt_prefix_index_single_scan`：

```sql
SET tidb_opt_prefix_index_single_scan = 'ON';
```

对于以下使用前缀索引的查询，从执行计划结果中可见，执行使用了索引 `idx_a_b` 且不需要回表。

```sql
mysql> EXPLAIN FORMAT='brief' SELECT COUNT(1) FROM t WHERE a = 1 AND b IS NOT NULL;
+--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
| id                       | estRows | task      | access object                | operator info                                         |
+--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
| StreamAgg                | 1.00    | root      |                              | funcs:count(Column#7)->Column#5                       |
| └─IndexReader            | 1.00    | root      |                              | index:StreamAgg                                       |
|   └─StreamAgg            | 1.00    | cop[tikv] |                              | funcs:count(1)->Column#7                              |
|     └─IndexRangeScan     | 99.90   | cop[tikv] | table:t, index:idx_a_b(a, b) | range:[1 -inf,1 +inf], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+------------------------------+-------------------------------------------------------+
4 rows in set (0.00 sec)
```
````

File: system-variables.md [hunk 6]
```diff
@@ -2364,6 +2364,69 @@ explain select * from t where age=5;
...
+### `tidb_opt_prefix_index_single_scan` <span class="version-mark">从 v6.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`ON`
+- 这个变量用于控制 TiDB 优化器是否将某些过滤条件下推到前缀索引，尽量避免不必要的回表，从而提高查询性能。
+- 将该变量设置为 `ON` 时，会将过滤条件下推到前缀索引。此时，假设一张表中 `col` 列是索引前缀列，查询语句中的 `col is null` 或者 `col is not null` 条件会被归为索引上的过滤条件，而不是回表时的过滤条件，从而避免不必要的回表。
+
+<details>
+<summary>该变量的使用示例如下：</summary>
```
Comment (TomShawn) [inline system-variables.md:2376]:
```suggestion
<summary>该变量的使用示例</summary>
```

**Comment by songrijie** [comment]
LGTM

File: system-variables.md [hunk 7]
```diff
@@ -2364,6 +2364,69 @@ explain select * from t where age=5;
...
+CREATE TABLE t (a INT, b VARCHAR(10), c INT, INDEX idx_a_b(a, b(5)));
+```
+
+此时关闭 `tidb_opt_prefix_index_single_scan`：
+
+```sql
+SET tidb_opt_prefix_index_single_scan = 'OFF';
+```
+
+对于以下查询，执行计划使用了前缀索引 `idx_a_b` 但需要回表（出现了 `TableRowIDScan` 算子）。
```
Comment (TomShawn) [inline system-variables.md:2390]:
```suggestion
对于以下查询，执行计划使用了前缀索引 `idx_a_b` 但需要回表（出现了 `IndexLookUp` 算子）。
```

**Review by TomShawn — APPROVED** [review]
rest LGTM

File: system-variables.md [hunk 8]
```diff
@@ -2364,6 +2364,69 @@ explain select * from t where age=5;
 - 范围：`[0, 1]`
 - 这个变量用来控制是否在 profile 输出中标记出对应的 SQL 语句，用于定位和排查性能问题。
 
+### `tidb_opt_prefix_index_single_scan` <span class="version-mark">从 v6.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`ON`
+- 这个变量用于控制 TiDB 优化器是否将某些过滤条件下推到前缀索引，尽量避免不必要的回表，从而提高查询性能。
+- 将该变量设置为 `ON` 时，会将过滤条件下推到前缀索引。此时，假设一张表中 `col` 列是索引前缀列，查询语句中的 `col is null` 或者 `col is not null` 条件会被归为索引上的过滤条件，而不是回表时的过滤条件，从而避免不必要的回表。
```
Comment (chrysan) [inline system-variables.md:2373]: 这里是否需要额外说明，不是打开就能避免回表，过滤条件和前缀索引需要满足一定条件。

Comment (hfxsd) [inline system-variables.md:2373]: 这句话中既有索引前缀、又有前缀所有，是否写错了？

Comment (xuyifangreeneyes) [reply system-variables.md:2373]: 前缀索引是指索引中某一列是 col(length) 的索引，索引前缀列是指索引中为 col(length) 形式的列。这里确实有些不太明确的地方。以下是 MySQL 里相关的一些描述：
https://dev.mysql.com/doc/refman/8.0/en/column-indexes.html#column-indexes-prefix
https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-column-prefixes

**Review by chrysan — APPROVED** [review]

**Pull request merged**